### PR TITLE
Implement separate context for arguments.

### PR DESCRIPTION
**Pull request by zherczeg**
#3683 should land first

### CODE REVIEW

File: jerry-core/parser/js/js-parser.c [hunk 1]
```diff
@@ -1867,17 +1899,39 @@ parser_parse_function_arguments (parser_context_t *context_p, /**< context */
...
+  if (context_p->status_flags & PARSER_LEXICAL_BLOCK_NEEDED)
+  {
+    if ((context_p->next_scanner_info_p->u8_arg & SCANNER_FUNCTION_LEXICAL_ENV_NEEDED)
+        || scanner_is_function_context_needed (context_p))
+    {
+      context_p->status_flags |= PARSER_LEXICAL_ENV_NEEDED;
+
+      parser_branch_t branch;
+      parser_emit_cbc_forward_branch (context_p, CBC_BLOCK_CREATE_CONTEXT, &branch);
+      parser_stack_push (context_p, &branch, sizeof(parser_branch_t));
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser.c:1920]: missing space before `(`;

File: jerry-core/parser/js/js-parser.c [hunk 2]
```diff
@@ -918,6 +918,30 @@ parser_post_processing (parser_context_t *context_p) /**< context */
...
+#ifndef JERRY_NDEBUG
+    PARSER_MINUS_EQUAL_U16 (context_p->context_stack_depth, PARSER_BLOCK_CONTEXT_STACK_ALLOCATION);
+#endif /* !JERRY_NDEBUG */
+
+    context_p->status_flags &= (uint32_t) ~PARSER_LEXICAL_BLOCK_NEEDED;
+
+    parser_emit_cbc (context_p, CBC_CONTEXT_END);
+
+    parser_branch_t branch;
+    parser_stack_pop (context_p, &branch, sizeof(parser_branch_t));
```
Comment (rerobika) [inline jerry-core/parser/js/js-parser.c:935]: ditto.

File: jerry-core/parser/js/js-scanner-util.c [hunk 3]
```diff
@@ -1588,36 +1596,31 @@ scanner_is_context_needed (parser_context_t *context_p) /**< context */
 #else /* !ENABLED (JERRY_ES2015) */
     JERRY_ASSERT (type == SCANNER_STREAM_TYPE_VAR);
 #endif /* ENABLED (JERRY_ES2015) */
+#endif /* !JERRY_NDEBUG */
 
-    size_t length;
-
-    if (!(data_p[0] & SCANNER_STREAM_UINT16_DIFF))
+    if (!(data & SCANNER_STREAM_UINT16_DIFF))
```
Comment (galpeter) [inline jerry-core/parser/js/js-scanner-util.c:1601]: Would it make sense to extract common parts? like:
```c
data_p += 2 + 1;
if (!(data & SCANNER_STREAM_UINT16_DIFF)))
{
  if (data_p[2] == 0)
  {
    data_p += sizeof (...);
  }
}
else
{
  data_p += 1;
}
```

Another question? Why "2" and "1" ? is there any info about these constants somewhere?

Comment (galpeter) [reply jerry-core/parser/js/js-scanner-util.c:1601]: I see this addition is also present in other parts. Maybe we could extact them into a single function to have them only at one place.

Comment (zherczeg) [reply jerry-core/parser/js/js-scanner-util.c:1601]: There is a long comment how the stream is compressed:
https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/parser/js/js-scanner.h#L114

I have been thinking about mergeng these functions, but I have doubts. They are similar, but they are also different. I don't know what is the best option here.

Comment (galpeter) [reply jerry-core/parser/js/js-scanner-util.c:1602]: What if we only create a function to calculate the required bytes to add? eg.:
```c
size_t scanner_stream_next_offset(uint8_t *data_p)
{
  size_t offset = 2 + 1;
  if (!(data_p[0] & SCANNER_STREAM_UINT16_DIFF)))
  {
    if (data_p[2] == 0)
    {
      offset += sizeof (...);
    }
  }
  else
  {
    offset+= 1;
  }
  return offset;
}
```

**Comment by zherczeg** [comment]
Fixes #3396.

File: jerry-core/parser/js/js-scanner-util.c (hunk 3 above)
Comment (zherczeg) [reply jerry-core/parser/js/js-scanner-util.c]: I decided to do the merge. I does not seems that bad, and this is not a hot code.

**Pull request merged**